File: datastructures/minheap/min_heap.go
package minheap

type MinHeap struct {
	array []int
}

func NewMinHeap() MinHeap {
	return MinHeap{
		array: make([]int, 0),
	}
}

func (h *MinHeap) Size() int {
	return len(h.array)
}

func (h *MinHeap) PeekMin() *int {
	if h.Size() == 0 {
		return nil
	}

	min := h.array[0]

	return &min
}

func (h *MinHeap) ExtractMin() *int {
	minVal := h.PeekMin()

	if minVal == nil {
		return nil
	}

	n := h.Size()
	h.array[0] = h.array[n]
	h.array = h.array[:n-1]
	h.bubbleDown(0)

	return minVal
}

func (h *MinHeap) Insert(key int) {
	h.array = append(h.array, key)
	h.bubbleUp(h.Size() - 1)
}

func (h *MinHeap) bubbleDown(index int) {
	indexSmallerChild := h.findSmallerChild(index)

	if indexSmallerChild >= h.Size() {
		return
	}

	if h.array[indexSmallerChild] < h.array[index] {
		h.array[indexSmallerChild], h.array[index] = h.array[index], h.array[indexSmallerChild]
		h.bubbleDown(indexSmallerChild)
	}
}

func (h *MinHeap) findSmallerChild(index int) (childIndex int) {
	size := h.Size()
	if index >= size {
		return -1
	}

	leftChildIndex, rightChildIndex := index*2+1, index*2+2
	if rightChildIndex >= size {
		if leftChildIndex >= size {
			return -1
		} else {
			return leftChildIndex
		}
	} else {
		if leftChildIndex < size && h.array[leftChildIndex] <= h.array[rightChildIndex] {
			return leftChildIndex
		} else {
			return rightChildIndex
		}
	}
}

func (h *MinHeap) bubbleUp(index int) {
	if index <= 0 {
		return
	}

	parentIndex := (index - 1) / 2
	if h.array[index] < h.array[parentIndex] {
		h.array[index], h.array[parentIndex] = h.array[parentIndex], h.array[index]
		h.bubbleUp(parentIndex)
	}
}

File: datastructures/minheap/min_heap.py
from __future__ import division

import sys


class MinHeap(object):
    def __init__(self):
        self.array = []

    def __len__(self):
        return len(self.array)

    def _find_smaller_child(self, index):
        left_child_index, right_child_index = 2 * index + 1, 2 * index + 2

        if right_child_index >= self.__len__():
            if left_child_index >= self.__len__():
                return -1
            else:
                return left_child_index
        else:
            if left_child_index < self.__len__() and self.array[left_child_index] <= self.array[right_child_index]:
                return left_child_index
            else:
                return right_child_index

    def _bubble_down(self, index):
        min_child_index = self._find_smaller_child(index)

        if min_child_index == -1:
            return

        if self.array[index] > self.array[min_child_index]:
            self.array[index], self.array[min_child_index] = self.array[min_child_index], self.array[index]
            self._bubble_down(min_child_index)

    def _bubble_up(self, index):
        if index == 0:
            return

        parent_index = (index - 1) // 2

        if self.array[index] < self.array[parent_index]:
            self.array[index], self.array[parent_index] = self.array[parent_index], self.array[index]
            self._bubble_up(parent_index)

    def peek_min(self):
        return self.array[0] if self.array else None

    def extract_min(self):
        if not self.array:
            return None

        if len(self.array) == 1:
            return self.array.pop(0)

        minVal = self.array[0]
        self.array[0] = self.array.pop(-1)
        self._bubble_down(index=0)

        return minVal

    def insert(self, key):
        if key is None:
            raise TypeError('key cannot be None')

        self.array.append(key)
        self._bubble_up(index=self.__len__() - 1)
